src/test/java/dinkplugin/util/RarityServiceTest.java

```diff
@@ -0,0 +1,262 @@
+
+    @Bind
+    private final Gson gson = RuneLiteAPI.GSON;
+
+    @Mock
+    @Bind
+    private ItemManager itemManager;
+
+    @Spy
+    @Bind
+    private RarityService service;
+
+    @Override
+    @BeforeEach
+    protected void setUp() {
+        super.setUp();
+
+        // default item mock
+        Mockito.doAnswer(invocation -> {
+            int id = invocation.getArgument(0);
+            ItemComposition comp = mock(ItemComposition.class);
+            when(comp.getMembersName()).thenReturn("?");
+            when(comp.getNote()).thenReturn(-1);
+            return comp;
+        }).when(itemManager).getItemComposition(anyInt());
+
+        // actual item mocks
+        mockItem(ItemID.DRAGON_SPEAR, "Dragon spear");
+        mockItem(ItemID.DRAGON_AXE, "Dragon axe");
+        mockItem(ItemID.SHARK, "Shark");
+        mockItem(ItemID.SNAPE_GRASS, "Snape grass");
+        mockItem(ItemID.GOBLIN_SKULL, "Goblin skull");
+        mockItem(ItemID.COINS, "Coins");
+        mockItem(ItemID.COINS_995, "Coins");
+        mockItem(ItemID.ARMADYL_HILT, "Armadyl hilt");
+        mockItem(ItemID.STEEL_ARROW, "Steel arrow");
+        mockItem(ItemID.SILVER_ORE, "Silver ore");
+        mockItem(ItemID.SILVER_ORE + 1, "Silver ore", true);
+        mockItem(ItemID.UNCUT_RUBY, "Uncut ruby");
+        mockItem(ItemID.ECUMENICAL_KEY, "Ecumenical key");
+        mockItem(ItemID.DEATH_RUNE, "Death rune");
+        mockItem(ItemID.SNAPE_GRASS_SEED, "Snape grass seed");
+        mockItem(ItemID.DRAGONSTONE, "Dragonstone");
+        mockItem(ItemID.ANCIENT_SHARD, "Ancient shard");
+        mockItem(ItemID.LARRANS_KEY, "Larran's key");
+        mockItem(ItemID.UNIDENTIFIED_RARE_FOSSIL, "Unidentified rare fossil");
+        mockItem(ItemID.GRIMY_GUAM_LEAF, "Grimy guam leaf");
+        mockItem(ItemID.DUST_BATTLESTAFF, "Dust battlestaff");
+        mockItem(ItemID.GRIMY_AVANTOE, "Grimy avantoe");
+        mockItem(ItemID.FIRE_TALISMAN, "Fire talisman");
+        mockItem(ItemID.CLUE_SCROLL_ELITE, "Clue scroll (elite)");
+        mockItem(ItemID.CLUE_SCROLL_ELITE_12157, "Clue scroll (elite)"); // wiki prefers this ID
+    }
+
+    @Test
+    @DisplayName("Ensure accurate rarity for very rare drops")
+    void testVeryRare() {
+        test("Aberrant spectre", ItemID.DRAGON_SPEAR, 1, 1.0 / 139_810);
+    }
+
+    @Test
+    @DisplayName("Ensure accurate rarity for rare drops")
+    void testRare() {
+        test("Dagannoth Supreme", ItemID.DRAGON_AXE, 1, 1.0 / 128);
+    }
+
+    @Test
+    @DisplayName("Ensure accurate rarity for uncommon drops")
+    void testUncommon() {
+        test("Dagannoth Supreme", ItemID.SHARK, 5, 5.0 / 128);
+    }
+
+    @Test
+    @DisplayName("Ensure accurate rarity for common drops")
+    void testCommon() {
+        test("Bree", ItemID.SNAPE_GRASS, 5, 7.0 / 127);
+    }
+
+    @Test
+    @Disabled("[BUG] doesn't distinguish by combat level: https://oldschool.runescape.wiki/w/Goblin#Drop_table_2")
```

Comment:
similar known bug: some scurrius drop rates are doubled since we don't distinguish between private vs. public kills https://oldschool.runescape.wiki/w/Scurrius

Reply:
similarly: green dragon gem drop table rates are doubled (level 79 vs. wildy slayer cave) https://oldschool.runescape.wiki/w/Green_dragon

Reply:
A possible mitigation: ignore duplicate drop entries (i.e., item id, quantity range, and rarity is same)

should address more edge cases than it introduces if yall agree @pajlada @Felanbird

Reply:
Done in https://github.com/pajlads/DinkPlugin/pull/425/commits/aa389a505de560d8bc1246408a594ac6a70eeca7